### PR TITLE
Improve error message on missing path segment keys

### PR DESCRIPTION
**Pull request by chamons**
Example:

```
   file:///C:/Users/chhamo/Programming/autorest.csharp/samples/CognitiveServices.TextAnalytics/foo.json#/components/schemas/paths┬╖v9pjtw┬╖v2-name-manifests-reference┬╖put┬╖responses┬╖201┬╖content┬╖application-json┬╖schema
FATAL: System.Collections.Generic.KeyNotFoundException:
Error while processing request /acr/v1/{name}/_manifests/{reference}: key `reference` was undefined.
Verify that it is correctly defined, and if not please file an issue at https://github.com/Azure/autorest.csharp/issues/new with a swagger that reproduces.
   at AutoRest.CSharp.Output.Models.RestClient.GetPathSegments(String httpRequestUri, Dictionary`2 parameters, Boolean isRaw)+MoveNext() in C:\Users\chhamo\Programming\autorest.csharp\src\AutoRest.CSharp\Output\Models\RestClient.cs:line 471
   at System.Collections.Generic.LargeArrayBuilder1.AddRange(IEnumerable1 items)
   at System.Collections.Generic.SparseArrayBuilder1.AddRange(IEnumerable1 items)
   at System.Collections.Generic.SparseArrayBuilder1.ReserveOrAdd(IEnumerable1 items)
   at System.Linq.Enumerable.Concat2Iterator`1.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
```

### CODE REVIEW

File: src/AutoRest.CSharp/AutoRest/Plugins/CSharpGen.cs [hunk 1]
```diff
@@ -134,7 +134,7 @@ public async Task<bool> Execute(IPluginCommunication autoRest)
             if (configuration.CredentialTypes.Contains("TokenCredential", StringComparer.OrdinalIgnoreCase) &&
                 configuration.CredentialScopes.Length < 1)
             {
-                await autoRest.Fatal("You are using TokenCredential wihtout passing in any credential-scopes.");
```
Comment (chamons) [inline src/AutoRest.CSharp/AutoRest/Plugins/CSharpGen.cs:137]: Just a random misspelling I saw, felt too trivial for another PR.

**Comment by chamons** [comment]
Hey @annelo-msft thoughts on this error message? Any tweaks you want to suggest while I'm under the hood here?

**Comment by annelo-msft** [comment]
Cool - thanks for doing this, @chamons!

Re: 
> key `reference` was undefined.

 
I think it's hard for me to make a leap here from what key is -- I know you're holding a dictionary of values somewhere, but I don't know what's in it or where it came from.  If it could tell me what block of JSON to look in because that's where the dictionary values are coming from, I think that would be more helpful.
 
For the error from yesterday, I'd love it if it could say:
 
> Error while processing request /acr/v1/{name}/_manifests/{reference}: `reference` not found in `parameters` block.
 
I'd give bonus points if it could tell me why it was looking for reference, e.g. 

> `reference` in URI path must have a matching definition in `parameters` block.
 

I like the part directing you to where to file an issue as well -- thanks for all this!

File: src/AutoRest.CSharp/Output/Models/RestClient.cs [hunk 2]
```diff
@@ -459,7 +459,18 @@ PathSegment TextSegment(string text)
...
-                yield return isLiteral ? TextSegment(text) : parameters[text];
+                if (isLiteral)
+                {
+                    yield return TextSegment(text);
+                }
+                else
+                {
+                    if (!parameters.ContainsKey (text))
+                    {
+                        throw new KeyNotFoundException ($"\n\nError while processing request {httpRequestUri}: key '{text}' was undefined.\nVerify that it is correctly defined, and if not please file an issue at https://github.com/Azure/autorest.csharp/issues/new with a swagger that reproduces.\n");
```
Comment (pakrym) [inline src/AutoRest.CSharp/Output/Models/RestClient.cs:470]: Let's be a bit more specific here. Something like "parameter is used in the path template but is not defined in the parameters collection"

**Review by pakrym — APPROVED** [review]

**Comment by chamons** [comment]
Next Text:
```
file:///C:/Users/chhamo/Programming/autorest.csharp/samples/CognitiveServices.TextAnalytics/foo.json#/components/schemas/paths┬╖v9pjtw┬╖v2-name-manifests-reference┬╖put┬╖responses┬╖201┬╖content┬╖application-json┬╖schema
FATAL: System.Collections.Generic.KeyNotFoundException:

Error while processing request /acr/v1/{name}/_manifests/{reference}
  Parameter 'reference' is used in the path template but is not defined in the parameters collection.
  'reference' in URI path must have a matching definition in parameters block.

   at AutoRest.CSharp.Output.Models.RestClient.GetPathSegments(String httpRequestUri, Dictionary`2 parameters, Boolean isRaw)+MoveNext() in C:\Users\chhamo\Programming\autorest.csharp\src\AutoRest.CSharp\Output\Models\RestClient.cs:line 472
   at System.Collections.Generic.LargeArrayBuilder1.AddRange(IEnumerable1 items)
```

**Comment by annelo-msft** [comment]
Small tweak suggestion:

>   Parameter 'reference' is used in the path template but is not defined in the parameters collection.

^ this is the POV of the C# code

>   'reference' in URI path must have a matching definition in parameters block.

^ this is the POV of the JSON swagger file

It might make sense to decide which POV you want to frame the explanation as being from.  Or do both, and clarify which is which.  i.e. in this one:

>   'reference' in URI path must have a matching definition in parameters block.

Perhaps consider adding the information "in the input OpenAPI JSON file," or something like this?


I'll also remark that the three sentences in this error refer to the same things in different ways:
- "Request", "path template" and "URI path"
- Parameter 'reference', 'reference'
- parameters collection, parameters block

Maybe just picking one for each to standardize on for each of these would help make it easier to understand.

**Comment by chamons** [comment]
I'm working on tweaking this, but I realized this isn't playing nice w\ my other PR that calls all exceptions as internal errors. I need to wrap or something.

**Comment by chamons** [comment]
Ok, what about this text @pakrym @annelo-msft :

```
Error while processing request '/acr/v1/{name}/_manifests/{reference}'

  'reference' in URI is missing a matching definition in the path parameters collection

Please review your swagger and make necessary corrections.

If the definition is correct file an issue at https://github.com/Azure/autorest.csharp/issues/new with details to reproduce.
```

**Comment by annelo-msft** [comment]
🌷 Ship it!

**Pull request merged**